Add note for array with empty struct or union (#365)
riscv-non-isa/riscv-elf-psabi-doc: riscv-cc.adoc

{ float f[1]; } g[2]; }` and `struct { float f; float g; }` are
treated the same.  Fields containing empty structs or unions are ignored while
flattening, even in {Cpp}, unless they have nontrivial copy constructors or

> destructors.  Fields containing zero-length bit-fields or zero-length array are

Reviewer: Grammatical suggestion:

```suggestion
destructors.  Fields containing zero-length bit-fields or zero-length arrays are
```

Reviewer 2: This was marked resolved but doesn't seem to have been integrated.

ignored while flattening.  Attributes such as `aligned` or `packed` do not
interfere with a struct's eligibility for being passed in registers according
to the rules below, i.e. `struct { int i; double d; }` and `+struct
__attribute__((__packed__)) { int i; double d }+` are treated the same, as are
`struct { float f; float g; }` and `+struct { float f; float g __attribute__
((aligned (8))); }+`.

> NOTE: One exceptional case for the flattening rule is the array with empty

Reviewer: Minor English suggestion:

```suggestion
NOTE: One exceptional case for the flattening rule is an array of empty
```

> structs or unions; C language treats it as an empty field, but {Cpp} language

Reviewer: Minor English suggestion:

```suggestion
structs or unions; C treats it as an empty field, but {Cpp}
```

treats it as a non-empty field since {Cpp} defines the size of an empty struct

> or union as 1. However, C defines it as 0, that result from the array with an
> empty struct or union having a different size between C and {Cpp}. i.e.

Reviewer: Delete redundant sentence and s/i.e./e.g.

```suggestion
or union as 1. E.g.,
```

> `struct { struct {} e[1]; float f; };` C will passed by `f0`, but {Cpp} will
> passed by `a0`
> 

Reviewer: `f0` is `ft0`: I think you mean `fa0`. 

Also, we're not using a fixed-width font for register mnemonics elsewhere, so I suggest not introducing it here and now.

Also a minor English suggestion.

```suggestion
for `struct { struct {} e[1]; float f; };`, C will pass `f` in fa0, but {Cpp} will
pass it in `a0`.
```

Reviewer: My experiment on LLVM for this example is showing the same result for C and C++. But having just embarrassed myself over at #366 I am very hesitant to say any more.

Reviewer 2: I opened #358 as it seemed Clang/LLVM wasn't matching GCC. My patch https://reviews.llvm.org/D142327 _should_ match the clarification in this PR but it hasn't been committed yet.

Reviewer: Thanks @asb . I think I understand how things work. I've updated my suggestion above to clarify some of my confusions.

treated the same.  Fields containing empty structs or unions are ignored while
flattening, even in {Cpp}, unless they have nontrivial copy constructors or
destructors.  Fields containing zero-length bit-fields or zero-length arrays are
ignored while flattening.  Attributes such as `aligned` or `packed` do not
interfere with a struct's eligibility for being passed in registers according
to the rules below, i.e. `struct { int i; double d; }` and `+struct
__attribute__((__packed__)) { int i; double d }+` are treated the same, as are
`struct { float f; float g; }` and `+struct { float f; float g __attribute__
((aligned (8))); }+`.

NOTE: One exceptional case for the flattening rule is an array of empty
structs or unions; C treats it as an empty field, but {Cpp}
treats it as a non-empty field since {Cpp} defines the size of an empty struct
or union as 1. i.e. for `struct { struct {} e[1]; float f; }` as the first
argument, C will treat it like `struct { float f; }` and pass `f` in `fa0` as
described below, whereas {Cpp} will pass the pass the entire aggregate in `a0` 
(XLEN = 64) or `a0` and `a1` (XLEN = 32), as described in the integer calling
convention.

> The zero-length arrays of empty structs or union will treat be

Reviewer 2: The zero-length => Zero-length
will treat be => will be

> ignored for bot C and {Cpp}. i.e. For `struct { struct {} e[0]; float f; };`,

Reviewer 2: bot => both
